### PR TITLE
Map extend

### PR DESCRIPTION
**Pull request by ZhangMan1126**
add pair gene
extend map with pair
tests for update

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt [hunk 1]
```diff
@@ -186,15 +200,57 @@ class MapGene<T>(
         elements.clear()
     }
 
-    fun removeElements(element: T){
+    /**
+     * remove [element] to [elements]
+     */
+    fun removeElements(element: PairGene<K, V>){
         elements.remove(element)
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt:207]: how does this work? doe genes implements `equalsTo`?  or is this assuming heap reference? add clarification

Comment (arcuri82) [reply core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt:207]: to clarify, how/where is this method supposed to be used?
if it is just a matter of removing an element, shouldn't just pass in a reference to the `K`?

File: core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt [hunk 2]
```diff
@@ -186,15 +200,57 @@ class MapGene<T>(
...
         element.removeThisFromItsBindingGenes()
     }
 
-    fun addElements(element: T){
+    /**
+     * add [element] to [elements]
+     */
+    fun addElements(element: PairGene<K, V>){
```
[the remark this answers is not in the document]
Comment (ZhangMan1126) [reply core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt:220]: there exist a method `containsKey` to check if the key exists. but now we only support integer, string and long type (see `getElementsBy` in map).
regarding the `addElements`, with a latest commit, now we handle it by 1) adding the new element if the key does not exist, or 2) replacing the existing one with the new element if the key exists;
regarding its mutation, we try to generate an element (key to value) with a key that does not exist in the current list of elements (see `addRandomElement`).

[the remark this answers is not in the document]
Comment (arcuri82) [reply core/src/main/kotlin/org/evomaster/core/search/gene/MapGene.kt:220]: i see. but, shouldn't it be renamed `addElement`? (ie without the plural `s` at the end)

**Review by arcuri82 — CHANGES_REQUESTED** [review]

**Pull request merged**